vtex/faststore
# add cookies functions

## What's the purpose of this pull request?

synchronize the order form of the vtex_session with the faststore's fs::cart.

## How it works?

Flow: https://miro.com/app/board/uXjVMe5dH6o=/

Session cookies are duplicated for the main domain and it must perform a validation and change the OrderForm to one synchronized with the external environment of the FS.

Add orderFormId param to my account and login page

## How to test it?

Use with this PR: https://github.com/vtex/admin-faststore/pull/16


3. Add a script to the checkout code:

```
function callUrlWithRetry(url, maxRetries, retryInterval) {
  return fetch(url,{
    method: 'POST',
  })
    .then(response => {
      if (response.ok) {
        return 
      } else {
        throw new Error(`Without vtex_session cookie`);
      }
    })
    .catch(error => {
      if (maxRetries > 0) {
        return new Promise(resolve => {
          setTimeout(() => {
            resolve(callUrlWithRetry(url, maxRetries - 1, retryInterval));
          }, retryInterval);
        });
      } else {
        throw error;
      }
    });
}
```

<img width="1167" alt="image" src="https://user-images.githubusercontent.com/67066494/223589533-2ab05250-bf92-41cd-b0ce-ecc418263480.png">


4. Now observe when accessing pages on IO the cookie being replicated to the main domain:

<img width="1167" alt="image" src="https://user-images.githubusercontent.com/67066494/223596641-809dcb99-f832-40f4-98cd-e9693f0a9ffc.png">

<img width="1167" alt="image" src="https://user-images.githubusercontent.com/67066494/223596748-5daa42d0-077b-49ce-a973-733e74cd6269.png">


With this PR, the order form of the session is validated in the method getsessionorder and if the order form is not the same as the one saved in the indexedDb, the change is made to synchronize the sessions.

Motivation:
- Reorder within my account
- Use Social Selling


https://user-images.githubusercontent.com/67066494/223597245-ccdcaf2d-e324-4bfd-bdc1-e05ef32f7aa8.mov



https://user-images.githubusercontent.com/67066494/223600066-ec2ef98f-b8c3-4bf8-8f0a-6d87a5188e3d.mov


## References

https://vtex.slack.com/archives/C0164SHMAV9/p1677609587367729


## TO DO

[ ] Store the session id inside indexedDB (fs::session)
[ ] Create tests using the vtex_session id context:  https://github.com/vtex/faststore/blob/daa79f6984f7a81ddb051ab0f45e4fb597a77142/packages/api/test/mutations.test.ts#L47
